res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+};
+
+PioneerDDJ400.tempoSliderLSB = function(channel, control, value, status, group) {
+    var fullValue = (PioneerDDJ400.highResMSB[group].tempoSlider << 7) + value;
+
+    engine.setValue(
+        group,
+        "rate",
+        ((0x4000 - fullValue) - 0x2000) / 0x2000
+    );
+};
+
+//
+// Effects
+//
+
+PioneerDDJ400.numFxSlots = 3;
+
+Object.defineProperty(PioneerDDJ400, "selectedFxSlot", {
+    get: function() {
+        return engine.getValue("[EffectRack1_EffectUnit1]", "focused_effect");
+    },
+    set: function(value) {
+        if (value < 0 || value > PioneerDDJ400.numFxSlots) {
+            return;
+        }
+        engine.setValue("[EffectRack1_EffectUnit1]", "focused_effect", value);
+        var isEffectEnabled = engine.getValue(PioneerDDJ400.selectedFxGroup, "enabled");
+        PioneerDDJ400.toggleLight(PioneerDDJ400.lights.beatFx, isEffectEnabled);
+    },
+});
+
+Object.defineProperty(PioneerDDJ400, "selectedFxGroup", {
+    get: function() {
+        return "[EffectRack1_EffectUnit1_Effect" + PioneerDDJ400.selectedFxSlot + "]";
+    },
+});
+
+PioneerDDJ400.beatFxLevelDepthRotate = function(_channel, _control, value) {
+    var newVal = value === 0 ? 0 : (value / 0x7F);
+
+    if (engine.getValue(PioneerDDJ400.selectedFxGroup, "enabled")) {
+        engine.softTakeoverIgnoreNextValue(PioneerDDJ400.selectedFxGroup, "meta");
+        engine.setValue("[EffectRack1_EffectUnit1]", "mix", newVal);
+    } else {
+        engine.softTakeoverIgnoreNextValue("[EffectRack1_EffectUnit1]", "mix");
+        engine.setValue(PioneerDDJ400.selectedFxGroup, "meta", newVal);
+    }
+};
+
+PioneerDDJ400.beatFxSelectPreviousEffect = function(_channel, _control, value) {
+    engine.setValue(PioneerDDJ400.selectedFxGroup, "prev_effect", value);
+};
+
+PioneerDDJ400.beatFxSelectNextEffect = function(_channel, _control, value) {
+    engine.setValue(PioneerDDJ400.selectedFxGroup, "next_effect", value);
+};
+
+PioneerDDJ400.beatFxLeftPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    PioneerDDJ400.selectedFxSlot = 1;
+};
+
+PioneerDDJ400.beatFxRightPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    PioneerDDJ400.selectedFxSlot = 2;
+};
+
+PioneerDDJ400.beatFxSelectPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    PioneerDDJ400.selectedFxSlot = 3;
+};
+
+PioneerDDJ400.beatFxOnOffPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    // toggle the currently focused effect slot in Effect Unit 1 (if any)
+    var selectedSlot = PioneerDDJ400.selectedFxSlot;
+    if (selectedSlot <= 0 || selectedSlot > PioneerDDJ400.numFxSlots) {
+        return;
+    }
+    var isEnabled = !engine.getValue(PioneerDDJ400.selectedFxGroup, "enabled");
+    engine.setValue(PioneerDDJ400.selectedFxGroup, "enabled", isEnabled);
+    PioneerDDJ400.toggleLight(PioneerDDJ400.lights.beatFx, isEnabled);
```

Comment:
The LED should be controlled by a callback connected to `enabled` or it will become out of sync if the user changes it from the GUI